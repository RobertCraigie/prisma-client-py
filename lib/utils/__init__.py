from ._utils import flatten as flatten
