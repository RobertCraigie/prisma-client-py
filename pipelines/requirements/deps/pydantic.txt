pydantic==2.3.0
