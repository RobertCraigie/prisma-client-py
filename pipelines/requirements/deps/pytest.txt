pytest==7.4.4
