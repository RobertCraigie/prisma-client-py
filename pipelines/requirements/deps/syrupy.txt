syrupy==4.6.4
