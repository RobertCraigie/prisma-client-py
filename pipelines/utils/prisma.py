from __future__ import annotations

import nox


def generate(
    session: nox.Session,
    *,
    schema: str | None = 'tests/data/schema.prisma',
    clean: bool = True,
) -> None:
    if clean:
        session.run('python', '-m', 'prisma_cleanup')

    if schema:
        args = (f'--schema={schema}',)
    else:
        args = ()

    session.run('prisma', 'generate', *args)
