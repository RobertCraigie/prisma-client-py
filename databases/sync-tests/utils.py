from __future__ import annotations

import os
from pydantic import BaseModel

from .._types import DatabaseMapping
from .._compat import LiteralString


CURRENT_DATABASE = os.environ['PRISMA_DATABASE']


class RawQueries(BaseModel):
    """Raw queries defined globally so they can be easily referenced in test functions"""

    count_posts: LiteralString

    find_user_by_id: LiteralString
    find_user_by_id_limit_1: LiteralString

    find_post_by_id: LiteralString
    find_posts_not_published: LiteralString

    select_unknown_table: LiteralString
    update_unique_post_title: LiteralString
    update_unique_post_new_title: LiteralString
    test_query_raw_no_result: LiteralString
    test_execute_raw_no_result: LiteralString


_mysql_queries = RawQueries(
    count_posts="""
        SELECT COUNT(*) as count
        FROM Post
    """,
    find_post_by_id="""
        SELECT *
        FROM Post
        WHERE id = ?
    """,
    find_user_by_id="""
        SELECT *
        FROM User
        WHERE User.id = ?
    """,
    find_user_by_id_limit_1="""
        SELECT *
        FROM User
        WHERE User.id = ?
        LIMIT 1
    """,
    select_unknown_table="""
        SELECT *
        FROM bad_table;
    """,
    find_posts_not_published="""
        SELECT id, published
        FROM Post
        WHERE published = false
    """,
    test_query_raw_no_result="""
        SELECT *
        FROM Post
        WHERE id = 'sdldsd'
    """,
    update_unique_post_title="""
        UPDATE Post
        SET title = 'My edited title'
        WHERE id = ?
    """,
    update_unique_post_new_title="""
        UPDATE Post
        SET title = 'My new title'
        WHERE id = ?
    """,
    test_execute_raw_no_result="""
        UPDATE Post
        SET title = 'updated title'
        WHERE id = 'sdldsd'
    """,
)

_postgresql_queries = RawQueries(
    count_posts="""
        SELECT COUNT(*) as count
        FROM "Post"
    """,
    find_post_by_id="""
        SELECT *
        FROM "Post"
        WHERE id = $1
    """,
    find_user_by_id="""
        SELECT *
        FROM "User"
        WHERE "User".id = $1
    """,
    find_user_by_id_limit_1="""
        SELECT *
        FROM "User"
        WHERE "User".id = $1
        LIMIT 1
    """,
    select_unknown_table="""
        SELECT *
        FROM bad_table;
    """,
    find_posts_not_published="""
        SELECT id, published
        FROM "Post"
        WHERE published = false
    """,
    test_query_raw_no_result="""
        SELECT *
        FROM "Post"
        WHERE id = 'sdldsd'
    """,
    update_unique_post_title="""
        UPDATE "Post"
        SET title = 'My edited title'
        WHERE id = $1
    """,
    update_unique_post_new_title="""
        UPDATE "Post"
        SET title = 'My new title'
        WHERE id = $1
    """,
    test_execute_raw_no_result="""
        UPDATE "Post"
        SET title = 'updated title'
        WHERE id = 'sdldsd'
    """,
)

RAW_QUERIES_MAPPING: DatabaseMapping[RawQueries] = {
    'postgresql': _postgresql_queries,
    'cockroachdb': _postgresql_queries,
    'mysql': _mysql_queries,
    'mariadb': _mysql_queries,
    'sqlite': RawQueries(
        count_posts="""
            SELECT COUNT(*) as count
            FROM Post
        """,
        find_post_by_id="""
            SELECT *
            FROM Post
            WHERE id = ?
        """,
        find_user_by_id="""
            SELECT *
            FROM User
            WHERE User.id = ?
        """,
        find_user_by_id_limit_1="""
            SELECT *
            FROM User
            WHERE User.id = ?
            LIMIT 1
        """,
        select_unknown_table="""
            SELECT *
            FROM bad_table;
        """,
        find_posts_not_published="""
            SELECT id, published
            FROM Post
            WHERE published = false
        """,
        test_query_raw_no_result="""
            SELECT *
            FROM Post
            WHERE id = 'sdldsd'
        """,
        update_unique_post_title="""
            UPDATE Post
            SET title = 'My edited title'
            WHERE id = ?
        """,
        update_unique_post_new_title="""
            UPDATE Post
            SET title = 'My new title'
            WHERE id = ?
        """,
        test_execute_raw_no_result="""
            UPDATE Post
            SET title = 'updated title'
            WHERE id = 'sdldsd'
        """,
    ),
}
