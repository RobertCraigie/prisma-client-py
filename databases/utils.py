from __future__ import annotations

import os
from typing import Set
from pathlib import Path
from typing_extensions import Literal, get_args, override

from pydantic import BaseModel
from syrupy.location import PyTestLocation
from syrupy.extensions.amber import AmberSnapshotExtension

from ._types import DatabaseMapping
from ._compat import LiteralString

DatabaseFeature = Literal[
    'enum',
    'json',
    'date',
    'arrays',
    'array_push',
    'json_arrays',
    'raw_queries',
    'create_many_skip_duplicates',
    'transactions',
    'case_sensitivity',
    'full_text_search',
]


class DatabaseConfig(BaseModel):
    id: str
    name: str
    env_var: str
    bools_are_ints: bool
    autoincrement_id: str
    unsupported_features: Set[DatabaseFeature]
    default_date_func: str

    # TODO: run this under coverage
    def supports_feature(self, feature: DatabaseFeature) -> bool:  # pragma: no cover
        if feature not in get_args(DatabaseFeature):
            raise RuntimeError(f'Unknown feature: {feature}')

        return feature not in self.unsupported_features


# ------------------ Test helpers ------------------

from .constants import TESTS_DIR, SYNC_TESTS_DIR

SHARED_SNAPSHOTS_DIR = Path(__file__).parent.joinpath('__shared_snapshots__')
CURRENT_DATABASE = os.environ.get('PRISMA_DATABASE')


class AmberSharedExtension(AmberSnapshotExtension):
    """Syrupy extension that stores the snapshots in a parent __shared_snapshots__ dir"""

    @classmethod
    @override
    def dirname(cls, *, test_location: PyTestLocation) -> str:
        test_dir = Path(test_location.filepath).parent
        if test_dir.is_relative_to(SYNC_TESTS_DIR):
            rel_dir = test_dir.relative_to(SYNC_TESTS_DIR)
        else:
            rel_dir = test_dir.relative_to(TESTS_DIR)

        return str(SHARED_SNAPSHOTS_DIR.joinpath(rel_dir))


class RawQueries(BaseModel):
    """Raw queries defined globally so they can be easily referenced in test functions"""

    count_posts: LiteralString

    find_user_by_id: LiteralString
    find_user_by_id_limit_1: LiteralString

    find_post_by_id: LiteralString
    find_posts_not_published: LiteralString

    select_unknown_table: LiteralString
    update_unique_post_title: LiteralString
    update_unique_post_new_title: LiteralString
    test_query_raw_no_result: LiteralString
    test_execute_raw_no_result: LiteralString


_mysql_queries = RawQueries(
    count_posts="""
        SELECT COUNT(*) as count
        FROM Post
    """,
    find_post_by_id="""
        SELECT *
        FROM Post
        WHERE id = ?
    """,
    find_user_by_id="""
        SELECT *
        FROM User
        WHERE User.id = ?
    """,
    find_user_by_id_limit_1="""
        SELECT *
        FROM User
        WHERE User.id = ?
        LIMIT 1
    """,
    select_unknown_table="""
        SELECT *
        FROM bad_table;
    """,
    find_posts_not_published="""
        SELECT id, published
        FROM Post
        WHERE published = false
    """,
    test_query_raw_no_result="""
        SELECT *
        FROM Post
        WHERE id = 'sdldsd'
    """,
    update_unique_post_title="""
        UPDATE Post
        SET title = 'My edited title'
        WHERE id = ?
    """,
    update_unique_post_new_title="""
        UPDATE Post
        SET title = 'My new title'
        WHERE id = ?
    """,
    test_execute_raw_no_result="""
        UPDATE Post
        SET title = 'updated title'
        WHERE id = 'sdldsd'
    """,
)

_postgresql_queries = RawQueries(
    count_posts="""
        SELECT COUNT(*) as count
        FROM "Post"
    """,
    find_post_by_id="""
        SELECT *
        FROM "Post"
        WHERE id = $1
    """,
    find_user_by_id="""
        SELECT *
        FROM "User"
        WHERE "User".id = $1
    """,
    find_user_by_id_limit_1="""
        SELECT *
        FROM "User"
        WHERE "User".id = $1
        LIMIT 1
    """,
    select_unknown_table="""
        SELECT *
        FROM bad_table;
    """,
    find_posts_not_published="""
        SELECT id, published
        FROM "Post"
        WHERE published = false
    """,
    test_query_raw_no_result="""
        SELECT *
        FROM "Post"
        WHERE id = 'sdldsd'
    """,
    update_unique_post_title="""
        UPDATE "Post"
        SET title = 'My edited title'
        WHERE id = $1
    """,
    update_unique_post_new_title="""
        UPDATE "Post"
        SET title = 'My new title'
        WHERE id = $1
    """,
    test_execute_raw_no_result="""
        UPDATE "Post"
        SET title = 'updated title'
        WHERE id = 'sdldsd'
    """,
)

RAW_QUERIES_MAPPING: DatabaseMapping[RawQueries] = {
    'postgresql': _postgresql_queries,
    'cockroachdb': _postgresql_queries,
    'mysql': _mysql_queries,
    'mariadb': _mysql_queries,
    'sqlite': RawQueries(
        count_posts="""
            SELECT COUNT(*) as count
            FROM Post
        """,
        find_post_by_id="""
            SELECT *
            FROM Post
            WHERE id = ?
        """,
        find_user_by_id="""
            SELECT *
            FROM User
            WHERE User.id = ?
        """,
        find_user_by_id_limit_1="""
            SELECT *
            FROM User
            WHERE User.id = ?
            LIMIT 1
        """,
        select_unknown_table="""
            SELECT *
            FROM bad_table;
        """,
        find_posts_not_published="""
            SELECT id, published
            FROM Post
            WHERE published = false
        """,
        test_query_raw_no_result="""
            SELECT *
            FROM Post
            WHERE id = 'sdldsd'
        """,
        update_unique_post_title="""
            UPDATE Post
            SET title = 'My edited title'
            WHERE id = ?
        """,
        update_unique_post_new_title="""
            UPDATE Post
            SET title = 'My new title'
            WHERE id = ?
        """,
        test_execute_raw_no_result="""
            UPDATE Post
            SET title = 'updated title'
            WHERE id = 'sdldsd'
        """,
    ),
}
