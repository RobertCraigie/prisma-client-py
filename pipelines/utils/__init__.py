from ._lark import generate_parsers as generate_parsers
from .utils import *
