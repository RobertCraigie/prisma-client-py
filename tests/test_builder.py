from __future__ import annotations

# fmt: off
# I prefer this way of formatting
import datetime
from typing import Any, Optional

import pytest
from pydantic import BaseModel
from syrupy.assertion import SnapshotAssertion

from prisma import models, PrismaMethod
from prisma.utils import _NoneType
from prisma.bases import _PrismaModel as PrismaModel
from prisma.builder import QueryBuilder, serializer
from prisma.errors import UnknownRelationalFieldError, UnknownModelError


# TODO: more tests
# TODO: cleanup registered serializers
# TODO: these tests should be schema agnostic
#       one schema should be used for all these tests
#       and only changed when needed for these tests
#       otherwise, simply adding a field to a model
#       will break these tests


def build_query(
    method: PrismaMethod,
    arguments: dict[str, Any],
    **kwargs: Any,
) -> str:
    return QueryBuilder(
        method=method,
        arguments=arguments,
        **kwargs,
    ).build_query()


def test_basic_building(snapshot: SnapshotAssertion) -> None:
    """Standard builder usage with and without a model"""
    query = QueryBuilder(
        method='find_unique',
        model=models.User,
        arguments={'where': {'id': '1'}}
    ).build_query()
    assert query == snapshot

    query = QueryBuilder(
        method='query_raw',
        arguments={'where': {'id': '1'}}
    ).build_query()
    assert query == snapshot


def test_invalid_include() -> None:
    """Invalid include field raises error"""
    with pytest.raises(UnknownRelationalFieldError) as exception:
        QueryBuilder(
            method='find_unique',
            model=models.User,
            arguments={
                'include': {
                    'hello': True,
                }
            }
        ).build_query()

    assert exception.match(
        'Field: "hello" either does not exist or is not a relational field on the User model'
    )


def test_include_no_model() -> None:
    """Trying to include a field without acess to a model raises an error"""
    with pytest.raises(ValueError) as exc:
        build_query(
            method='query_raw',
            arguments={'include': {'posts': True}}
        )

    assert exc.match('Cannot include fields when model is None.')


def test_include_with_arguments(snapshot: SnapshotAssertion) -> None:
    """Including a field with filters"""
    query = QueryBuilder(
        method='find_unique',
        model=models.User,
        arguments={
            'where': {'id': 1},
            'include': {'posts': {'where': {'id': 1}}}
        }
    ).build_query()
    assert query == snapshot


def test_raw_queries(snapshot: SnapshotAssertion) -> None:
    """Raw queries serialise paramaters to JSON"""
    query = QueryBuilder(
        method='query_raw',
        arguments={
            'query': 'SELECT * FROM User where id = $1',
            'parameters': ["1263526"],
        }
    ).build_query()
    assert query == snapshot


def test_datetime_serialization_tz_aware(snapshot: SnapshotAssertion) -> None:
    """Serializing a timezone aware datetime converts to UTC"""
    query = QueryBuilder(
        method='find_unique',
        model=models.Post,
        arguments={
            'where': {
                'created_at': datetime.datetime(1985, 10, 26, 1, 1, 1, tzinfo=datetime.timezone.max)
            }
        }
    ).build_query()
    assert query == snapshot


def test_datetime_serialization_tz_unaware(snapshot: SnapshotAssertion) -> None:
    """Serializing a timezone naive datetime converts to UTC"""
    query = QueryBuilder(
        method='find_unique',
        model=models.Post,
        arguments={
            'where': {
                'created_at': datetime.datetime(1985, 10, 26, 1, 1, 1)
            }
        }
    ).build_query()
    assert query == snapshot


def test_unicode(snapshot: SnapshotAssertion) -> None:
    """Serializing unicode strings does not convert to ASCII"""
    query = QueryBuilder(
        method='find_unique',
        model=models.User,
        arguments={
            'where': {
                'name': '❤',
            }
        }
    ).build_query()
    assert query == snapshot


def test_unknown_model() -> None:
    """Passing unknown model raises an error"""
    class FooModel(PrismaModel):
        __prisma_model__ = 'Foo'

    with pytest.raises(UnknownModelError) as exc:
        QueryBuilder(
            method='find_unique',
            model=FooModel,
            arguments={},
        ).build_query()

    assert exc.match(r'Model: "Foo" does not exist\.')


def test_unserializable_type() -> None:
    """Passing an unserializable type raises an error"""
    with pytest.raises(TypeError) as exc:
        QueryBuilder(
            method='find_first',
            arguments={
                'where': QueryBuilder
            }
        ).build_query()

    assert exc.match(r'Type <class \'prisma.builder.QueryBuilder\'> not serializable')


def test_unserializable_instance() -> None:
    """Passing an unserializable instance raises an error"""
    with pytest.raises(TypeError) as exc:
        QueryBuilder(
            method='find_first',
            arguments={
                'where': _NoneType()
            }
        ).build_query()

    assert exc.match(r'Type <class \'prisma.utils._NoneType\'> not serializable')


def test_custom_serialization(snapshot: SnapshotAssertion) -> None:
    """Registering a custom serializer serializes as expected"""
    class Foo:
        def __init__(self, arg: int) -> None:
            self.arg = arg

    @serializer.register(Foo)
    def custom_serializer(inst: Foo) -> int:  # pyright: ignore[reportUnusedFunction]
        return inst.arg

    query = QueryBuilder(
        method='find_unique',
        model=models.Post,
        arguments={
            'where': {
                'title': Foo(1),
            }
        }
    ).build_query()
    assert query == snapshot


def test_select(snapshot: SnapshotAssertion) -> None:
    """Selecting a subset of fields"""
    class OtherModel(PrismaModel):
        name: str
        __prisma_model__ = 'User'

    class CustomModel(PrismaModel):
        published: bool
        author: Optional[OtherModel]

        __prisma_model__ = 'Post'

    CustomModel.update_forward_refs(**locals())

    query = QueryBuilder(
        method='find_first',
        model=CustomModel,
        arguments={
            'where': {
                'title': 'Foo',
            },
        },
    ).build_query()
    assert query == snapshot

    with pytest.raises(UnknownRelationalFieldError) as exc:
        QueryBuilder(
            method='find_unique',
            model=OtherModel,
            arguments={
                'include': {
                    'posts': True,
                },
            },
        ).build_query()

    assert exc.match(r'Field: "posts" either does not exist or is not a relational field on the OtherModel model')

    query = QueryBuilder(
        method='find_first',
        model=CustomModel,
        arguments={
            'include': {
                'author': True,
            },
        },
    ).build_query()
    assert query == snapshot


def test_select_non_prisma_model_basemodel(snapshot: SnapshotAssertion) -> None:
    """Fields that point to a `BaseModel` but do not set the `__prisma_model__`
    class variable are included by default as scalar fields.
    """
    class OtherModel(PrismaModel):
        name: str
        __prisma_model__ = 'User'

    class TypedJson(BaseModel):
        foo: str

    class CustomModel(PrismaModel):
        published: bool
        my_json_blob: TypedJson
        author: Optional[OtherModel]

        __prisma_model__ = 'Post'

    CustomModel.update_forward_refs(**locals())

    query = QueryBuilder(
        method='find_first',
        model=CustomModel,
        arguments={
            'where': {
                'title': 'Foo',
            },
        },
    ).build_query()
    assert query == snapshot


def test_forward_reference_error() -> None:
    """If a Model has forward references and doesn't call `update_forward_refs()` then we can't
    tell if a given field should be included by default. Hence the sanity check here.
    """

    class TypedJson(BaseModel):
        foo: str

    class CustomModel(PrismaModel):
        meta: TypedJson

        __prisma_model__ = 'Post'

    with pytest.raises(RuntimeError, match=r'Forward references must be evaluated using CustomModel\.update_forward_refs\(\)'):
        QueryBuilder(
            method='find_first',
            model=CustomModel,
            arguments={},
        ).build_query()
