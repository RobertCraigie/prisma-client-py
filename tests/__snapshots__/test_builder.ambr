# name: test_basic_building
  '''
  query {
    result: findUniqueUser
    (
      where: {
        id: "1"
      }
    )
    {
      id
      name
      email
      created_at
    }
  }
  '''
# ---
# name: test_basic_building.1
  '''
  mutation {
    result: queryRaw
    (
      where: {
        id: "1"
      }
    )
  }
  '''
# ---
# name: test_custom_serialization
  '''
  query {
    result: findUniquePost
    (
      where: {
        title: 1
      }
    )
    {
      id
      created_at
      updated_at
      title
      published
      views
      desc
      author_id
    }
  }
  '''
# ---
# name: test_datetime_serialization_tz_aware
  '''
  query {
    result: findUniquePost
    (
      where: {
        created_at: "1985-10-25T01:02:01+00:00"
      }
    )
    {
      id
      created_at
      updated_at
      title
      published
      views
      desc
      author_id
    }
  }
  '''
# ---
# name: test_datetime_serialization_tz_unaware
  '''
  query {
    result: findUniquePost
    (
      where: {
        created_at: "1985-10-26T01:01:01+00:00"
      }
    )
    {
      id
      created_at
      updated_at
      title
      published
      views
      desc
      author_id
    }
  }
  '''
# ---
# name: test_include_with_arguments
  '''
  query {
    result: findUniqueUser
    (
      where: {
        id: 1
      }
    )
    {
      id
      name
      email
      created_at
      posts(
        where: {
          id: 1
        }
      )
      {
        id
        created_at
        updated_at
        title
        published
        views
        desc
        author_id
      }
    }
  }
  '''
# ---
# name: test_raw_queries
  '''
  mutation {
    result: queryRaw
    (
      query: "SELECT * FROM User where id = $1"
      parameters: "[\"1263526\"]"
    )
  }
  '''
# ---
# name: test_select
  '''
  query {
    result: findFirstPost
    (
      where: {
        title: "Foo"
      }
    )
    {
      published
    }
  }
  '''
# ---
# name: test_select.1
  '''
  query {
    result: findFirstPost
    {
      published
      author {
        name
      }
    }
  }
  '''
# ---
# name: test_select_non_prisma_model_basemodel
  '''
  query {
    result: findFirstPost
    (
      where: {
        title: "Foo"
      }
    )
    {
      published
      my_json_blob
    }
  }
  '''
# ---
# name: test_unicode
  '''
  query {
    result: findUniqueUser
    (
      where: {
        name: "‚ù§"
      }
    )
    {
      id
      name
      email
      created_at
    }
  }
  '''
# ---
